### ISSUE TITLE
Incremental builds fail with obscure ocamldep error

### ISSUE THREAD

**Issue by snowleopard** [ocaml/dune#5064]
When working on `build_system.ml` in file-watching mode I occasionally get myself into a bad state where Dune fails with an error like this:

```
Error: ocamldep returned unexpected output for
_build/default/src/dune_engine/build_system.ml:

-> required by
   _build/default/src/dune_engine/.dune_engine.objs/dune_engine__Build_system.impl.all-deps
-> required by
   _build/default/src/dune_engine/.dune_engine.objs/native/dune_engine__Build_system.cmx
-> required by _build/default/src/dune_engine/dune_engine.cmxa
-> required by
   _build/default/test/unit-tests/artifact_substitution/artifact_substitution.exe
-> required by alias test/unit-tests/artifact_substitution/runtest in
   test/unit-tests/artifact_substitution/dune:2
```

I don't have a reliable reproducer but I believe this happens when the set of dependencies of the `build_system.ml` changes, e.g. if I start/stop depending on another module.

The only way to get myself out of this bad state is to `rm -rf _build` and then build with `--cache=disabled`. So, my understanding is that our ocamldep rules are somehow non-deterministic, probably only in the file-watching mode as otherwise this would have been noticed long ago.

I wonder if it's related to stateful code in Stdune's `top_closure.ml`.

@rgrinberg Have you come across anything like this?

**Comment by rgrinberg** [comment]
> So, my understanding is that our ocamldep rules are somehow non-deterministic, probably only in the file-watching mode as otherwise this would have been noticed long ago.

I'm pretty sure they are deterministic. At least we make them deterministic by sorting the results.

Somehow ocamldep is returning us an empty file as the output. I'm wondering if it's a race condition of some sorts? We're observing the ocamldep output before its fully written? I would try sandboxing the ocamldep rules and seeing if that helps.

> I wonder if it's related to stateful code in Stdune's top_closure.ml.

No idea, but it seems to be worth fixing regardless.

**Comment by snowleopard** [comment]
I believe this was fixed in #5128.
